plugins.go

```diff
@@ -116,16 +107,98 @@ func findAndLoadPlugins(o *Options) error {
 	return nil
 }
 
+func (o *Options) LoadMultiPlugins(found map[string]string) error {
+	for _, plug := range o.MultiPlugins {
+		name := plug[0]
+		path, ok := found[name]
+		if !ok {
+			return fmt.Errorf("mutlitype plugin %s not found in plugin dirs\n", name)
+		}
+		fltr, pred, dc, err := LoadMultiPlugin(path, plug[1:])
+		if err != nil {
+			return fmt.Errorf("failed to load plugin %s: %s\n", path, err)
+		}
+
+		if fltr != nil {
+			o.CustomFilters = append(o.CustomFilters, fltr)
+		}
+		if pred != nil {
+			o.CustomPredicates = append(o.CustomPredicates, pred)
+		}
+		if dc != nil {
+			o.CustomDataClients = append(o.CustomDataClients, dc)
+		}
+		fmt.Printf("mutlitype plugin %s loaded from %s (filter: %t, predicate: %t, dataclient: %t)\n",
+			name, path, fltr != nil, pred != nil, dc != nil)
+		delete(found, name)
+	}
+	return nil
+}
+
+func LoadMultiPlugin(path string, args []string) (filters.Spec, routing.PredicateSpec, routing.DataClient, error) {
+	mod, err := plugin.Open(path)
+	if err != nil {
+		return nil, nil, nil, fmt.Errorf("open multitype plugin %s: %s", path, err)
+	}
+
+	conf, err := readPluginConfig(path)
+	if err != nil {
+		return nil, nil, nil, fmt.Errorf("failed to read config for %s: %s", path, err)
+	}
+
+	sym, err := mod.Lookup("InitPlugin")
+	if err != nil {
+		return nil, nil, nil, fmt.Errorf("lookup module symbol failed for %s: %s", path, err)
+	}
+	return loadMultiPlugin(sym, path, append(conf, args...))
+}
+
+func loadMultiPlugin(sym plugin.Symbol, path string, args []string) (filters.Spec, routing.PredicateSpec, routing.DataClient, error) {
+	fn, ok := sym.(func([]string) (filters.Spec, routing.PredicateSpec, routing.DataClient, error))
```

Comment:
How would it be to have an `InitPlugin(opts []string) ([]filters.Spec, []routing.Predicate, []routing.DataClient)` generic method with slices of return values, that can return multiple of any filters, predicates and data clients?  This could be the generic plugin interface, while we could still keep the InitFilter, InitPredicate and InitDataClient interfaces to support the implementations of simpler plugins, with a single return value.